File: java/src/main/java/com/ngt/state/ListStateDemo02.java
package com.ngt.state;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.restartstrategy.RestartStrategies;
import org.apache.flink.api.common.state.ListState;
import org.apache.flink.api.common.state.ListStateDescriptor;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.time.Time;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

import java.util.ArrayList;
import java.util.List;

/**
 * @author ngt
 * @create 2021-01-31 3:09
 * 不使用List，而是使用value 实现list状态保存的功能
 */
public class ListStateDemo02 {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        DataStreamSource<String> lines = env.socketTextStream("192.168.31.8", 8888);

        env.enableCheckpointing(5000);
        // 未设置重启策略的时候程序出现异常就会退出，设置重启策略
        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(3, Time.seconds(5)));

        SingleOutputStreamOperator<Tuple2<String, String>> tpDataStream = lines.map(new MapFunction<String, Tuple2<String, String>>() {
            @Override
            public Tuple2<String, String> map(String value) throws Exception {
                String[] words = value.split(",");
                return Tuple2.of(words[0], words[1]);
            }
        });

        KeyedStream<Tuple2<String, String>, String> keyedStream = tpDataStream.keyBy(d -> d.f0);

        keyedStream.process(new KeyedProcessFunction<String, Tuple2<String, String>, Tuple2<String, List<String>>>() {
            private transient ValueState<List<String>> listState;

            @Override
            public void open(Configuration parameters) throws Exception {
                ValueStateDescriptor<List<String>> stateDescriptor =
                        new ValueStateDescriptor<>("wc", TypeInformation.of(new TypeHint<List<String>>() {
                        }));
                listState = getRuntimeContext().getState(stateDescriptor);
            }

            @Override
            public void processElement(Tuple2<String, String> value, Context ctx, Collector<Tuple2<String, List<String>>> out) throws Exception {
                String action = value.f1;
                List<String> lst = listState.value();
                if (lst == null) {
                    lst = new ArrayList<String>();
                }
                lst.add(action);
                //更新状态
                listState.update(lst);
                out.collect(Tuple2.of(value.f0,lst));
            }
        }).print();
        env.execute();
    }
}

File: scala/src/main/scala/com/ngt/state/ListStateDemo02.scala
package com.ngt.state

import org.apache.flink.api.common.restartstrategy.RestartStrategies
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.api.common.time.Time
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

import scala.collection.mutable.ListBuffer

/**
 * @author ngt
 * @create 2021-02-08 21:43
 */
object ListStateDemo02 {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    val lines: DataStream[String] = env.socketTextStream("192.168.31.8", 8888)

    // 启用 Checkpointing
    env.enableCheckpointing(5000)
    // 未设置重启策略的时候程序出现异常就会退出，设置重启策略
    env.setRestartStrategy(RestartStrategies.fixedDelayRestart(3, Time.seconds(5)))

    val tpDataStream: DataStream[(String, String)] = lines.map(data => {
      val words: Array[String] = data.split(",")
      (words(0), words(1))
    })

    tpDataStream.keyBy(_._1)
      .process(new KeyedProcessFunction[String, (String, String), (String, ListBuffer[String])] {
        // scala 中的状态可以使用 lazy，也可以使用 open，使用open的时候和 Java 写法一样

        var listState: ValueState[ListBuffer[String]] = _

        override def open(parameters: Configuration): Unit = {
          // 注意非基本类型要使用  TypeInformation.of(classOf[ListBuffer[String]])
          val stateDescriptor: ValueStateDescriptor[ListBuffer[String]] = new ValueStateDescriptor[ListBuffer[String]]("wc", TypeInformation.of(classOf[ListBuffer[String]]))
          listState = getRuntimeContext.getState(stateDescriptor)
        }

        override def processElement(value: (String, String),
                                    ctx: KeyedProcessFunction[String, (String, String), (String, ListBuffer[String])]#Context, out: Collector[(String, ListBuffer[String])]): Unit = {
          val action: String = value._2
          var lst: ListBuffer[String] = listState.value()
          if (lst == null) {
            lst = new ListBuffer[String]
          }
          lst += action
          listState.update(lst)
          out.collect((value._1, lst))
        }
      })
      .print()

    env.execute()
  }
}

/*
user1,A
user1,B
user2,D
user2,C
ueer1,C

1> (user2,ListBuffer(C))
6> (user1,ListBuffer(A))
1> (user2,ListBuffer(C, D))
6> (user1,ListBuffer(A, B))
7> (ueer1,ListBuffer(C))
 */